### PR TITLE
feature/refactoring cafe list

### PR DESCRIPTION
**Pull request by xidealo**

### CODE REVIEW

File: app/src/main/java/com/bunbeauty/papakarlo/feature/cafe/screen/cafelist/CafeListFragment.kt [hunk 1]
```diff
@@ -101,78 +97,79 @@ class CafeListFragment : BaseFragmentWithSharedViewModel(R.layout.layout_compose
...
-                CafeItem(
-                    modifier = Modifier,
-                    cafeItem = cafeItem,
-                    onClick = {
-                        onCafeClicked(cafeItem.uuid)
-                    }
-                )
+                FoodDeliveryItem(needDivider = !cafeItem.isLast) {
+                    CafeItem(
+                        modifier = Modifier,
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/cafe/screen/cafelist/CafeListFragment.kt:108]: extra

File: app/src/main/java/com/bunbeauty/papakarlo/feature/cafe/screen/cafelist/CafeListViewState.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package com.bunbeauty.papakarlo.feature.cafe.screen.cafelist
+
+import com.bunbeauty.papakarlo.feature.cafe.ui.CafeItemAndroid
+import com.bunbeauty.papakarlo.feature.topcart.TopCartUi
+import com.bunbeauty.shared.presentation.base.BaseViewState
+
+data class CafeListViewState(
```
Comment (MaxAstin) [inline app/src/main/java/com/bunbeauty/papakarlo/feature/cafe/screen/cafelist/CafeListViewState.kt:7]: @Immutable

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt [hunk 3]
```diff
@@ -1,79 +1,96 @@
...
+                            copy(
+                                cafeList = cafeWithOpenStateList.mapIndexed { index, cafeWithOpenState ->
+                                    cafeWithOpenState.toCafeItem(
+                                        isLast = index == cafeWithOpenStateList.lastIndex
+                                    )
+                                },
+                                isLoading = false
+                            )
+                        }
+                    }.launchIn(sharedScope)
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt:85]: Already in `sharedScope` 
Lets just use `colect`

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt [hunk 4]
```diff
@@ -1,79 +1,96 @@
...
-                    oldState.copy(
-                        cafeList = cafeWithOpenStateList.map { cafeWithOpenState ->
-                            cafeWithOpenState.toCafeItem()
-                        },
-                        state = CafeListState.State.Success
-                    )
+    private fun observeCart() {
+        sharedScope.launchSafe(
+            block = {
+                observeCartUseCase().collectLatest { cartTotalAndCount ->
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt:55]: Let's cancel previous job

File: shared/build.gradle.kts [hunk 5]
```diff
@@ -117,6 +117,10 @@ android {
         release {}
     }
 
+    buildFeatures {
```
Comment (MaxAstin) [inline shared/build.gradle.kts:120]: Do we need this?

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt (hunk 3 above)
Comment (xidealo) [reply shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt:85]: we need job
How to get it with collect?

Comment (xidealo) [reply shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt:85]: need check

File: shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt [hunk 6]
```diff
@@ -1,79 +1,103 @@
...
+) : SharedStateViewModel<CafeList.DataState, CafeList.Action, CafeList.Event>(
+    initDataState = CafeList.DataState(
+        cafeList = listOf(),
+        cartCostAndCount = null,
+        isLoading = true
+    )
+) {
 
     private var observeCafeListJob: Job? = null
+    private var observeCart: Job? = null
```
Comment (MaxAstin) [inline shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt:30]: `observeCartJob`

**Review by MaxAstin — APPROVED** [review]

**Pull request merged**